domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt

```diff
@@ -735,4 +743,164 @@ class ExplorationProgressController @Inject constructor(
+   * created.
+   *
+   * @return the pending [State] for the current exploration
+   */
+  private fun createPendingTopStateFromCheckpoint(progress: ExplorationProgress): State {
+    val pendingTopState =
+      progress.stateGraph.getState(progress.currentExplorationCheckpoint.pendingStateName)
+    val hintList = createHintListFromCheckpoint(
+      pendingTopState.interaction.hintList,
+      progress.currentExplorationCheckpoint.helpIndex
+    )
+    val solution = createSolutionFromCheckpoint(pendingTopState, progress.hintState.helpIndex)
+    val interactionBuilder =
+      pendingTopState.interaction.toBuilder()
+        .clearHint()
+        .addAllHint(hintList)
+        .setSolution(solution)
+        .build()
+    return pendingTopState.toBuilder().setInteraction(interactionBuilder).build()
+  }
+
+  /**
+   * Mark all hints as reveled in the pendingState that were revealed for the current state pending
+   * state before the checkpoint was saved.
+   *
+   * @param pendingStateHintList the list of hint for the current pending state
+   * @param helpIndex the state of hints for the exploration which was generated using the saved
+   *     checkpoint
+   */
+  private fun createHintListFromCheckpoint(
+    pendingStateHintList: List<Hint>,
+    helpIndex: HelpIndex
+  ): List<Hint> {
+    val updatedHintList: MutableList<Hint> = ArrayList()
+    when (helpIndex.indexTypeCase) {
+      HelpIndex.IndexTypeCase.HINT_INDEX -> {
+        pendingStateHintList.forEachIndexed { index, hint ->
+          if (index < helpIndex.hintIndex) {
+            // Mark all hints as visible and revealed which have an index less than that stored in
+            // the HintState.
+            updatedHintList.add(hint.toBuilder().setHintIsRevealed(true).build())
+          } else if (index == helpIndex.hintIndex) {
+            // Add the currently shown hint to the updated hint list and set hintIsRevealed
+            // depending upon if the current hint was revealed or not.
+            updatedHintList.add(
+              hint.toBuilder().setHintIsRevealed(helpIndex.isAllVisibleHelpRevealed).build()
+            )
+          } else {
+            // Add all the remaining hints that are not yet visible to the user to the updated hint
+            // list.
+            updatedHintList.add(
+              hint.toBuilder().setHintIsRevealed(false).build()
+            )
+          }
+        }
+      }
+      HelpIndex.IndexTypeCase.SHOW_SOLUTION, HelpIndex.IndexTypeCase.EVERYTHING_REVEALED -> {
+        // All the hints are visible and revealed if helpIndex.indexTypeCase is equal to
+        // SHOW_SOLUTION or EVERYTHING_REVEALED.
+        pendingStateHintList.forEach { hint ->
+          updatedHintList.add(hint.toBuilder().setHintIsRevealed(true).build())
+        }
+      }
+      else -> updatedHintList.addAll(pendingStateHintList)
+    }
+    return updatedHintList
```

Comment:
```suggestion
    return when (helpIndex.indexTypeCase) {
      HelpIndex.IndexTypeCase.HINT_INDEX -> {
        pendingStateHintList.forEachIndexed { index, hint ->
          if (index < helpIndex.hintIndex) {
            // Mark all hints as visible and revealed which have an index less than that stored in
            // the HintState.
            listOf(hint.toBuilder().setHintIsRevealed(true).build())
          } else if (index == helpIndex.hintIndex) {
            // Add the currently shown hint to the updated hint list and set hintIsRevealed
            // depending upon if the current hint was revealed or not.
            listOf(hint.toBuilder().setHintIsRevealed(helpIndex.isAllVisibleHelpRevealed).build())
          } else {
            // Add all the remaining hints that are not yet visible to the user to the updated hint
            // list.
            listOf(hint.toBuilder().setHintIsRevealed(false).build())
          }
        }
      }
      HelpIndex.IndexTypeCase.SHOW_SOLUTION, HelpIndex.IndexTypeCase.EVERYTHING_REVEALED -> {
        // All the hints are visible and revealed if helpIndex.indexTypeCase is equal to
        // SHOW_SOLUTION or EVERYTHING_REVEALED.
        pendingStateHintList.forEach { hint ->
          listOf(hint.toBuilder().setHintIsRevealed(true).build())
        }
      }
      else -> pendingStateHintList
    }
```

Better to avoid mutable state. Ditto elsewhere.

Reply:
changed to mapIndexed